### PR TITLE
Use Expr expression in RelAlg nodes instead of Rex

### PR DESCRIPTION
**Pull request by ienkovich**
This is only a partial replacement of Rex with Exprs. Included changes:
 - Moves `Expr` classes to a separate lib and into `hdk::ir` namespace
 - Add `Expr` expressions in addition to `Rex` expression to all `RelAlgNode` classes
 - Use added `Expr`s for execution units creation. Existing `Rex` expressions are used for debugging purposes to check `Rex` and `Expr` variants match
 - Some parts of code analysis (visitor based mostly) moved to `Expr` usage

Right now I'm working on completely removing `Rex` expressions from `RelAlgNode` classes and moving the rest of the code analysis and transformations to `Expr`s.

### CODE REVIEW

**Comment by alexbaden** [comment]
Looks reasonable. Can we split the first 4 commits and merge them now? Will make the change set smaller I suspect and reduce conflicts.

**Comment by ienkovich** [comment]
> Looks reasonable. Can we split the first 4 commits and merge them now? Will make the change set smaller I suspect and reduce conflicts.

Yes, I'll try to split this PR.

File: IR/Expr.cpp [hunk 1]
```diff
@@ -219,6 +219,18 @@ bool Datum_equal(const SQLTypeInfo& ti, Datum val1, Datum val2) {
 
 }  // namespace
 
+void OrderEntry::print() const {
```
Comment (kurapov-peter) [inline IR/Expr.cpp:222]: Are these for debug only? Why not << operator?

**Review by kurapov-peter — COMMENTED** [review]
I think I went through most of it...

File: QueryEngine/RelAlgDagBuilder.cpp [hunk 2]
```diff
@@ -54,7 +55,7 @@ void RexSubQuery::setExecutionResult(
 }
 
 std::unique_ptr<RexSubQuery> RexSubQuery::deepCopy() const {
-  return std::make_unique<RexSubQuery>(type_, result_, ra_->deepCopy());
+  return std::make_unique<RexSubQuery>(type_, result_, ra_);
```
[the remark this answers is not in the document]
Comment (ienkovich) [reply QueryEngine/RelAlgDagBuilder.cpp:58]: I don't know what was the original idea here. I see no reason for copying the underlying node when `RexSubQuery` is copied. And these copies made it impossible to use the same node in the corresponding `ScalarSubquery` expression which really messed up the transition. So, I simply removed the copy and nothing wrong happened.

File: QueryEngine/RelAlgDagBuilder.cpp [hunk 3]
```diff
@@ -1092,19 +1200,17 @@ JoinType to_join_type(const std::string& join_type_name) {
...
     const RANodeOutput& ra_output) noexcept {
   std::vector<std::unique_ptr<const RexScalar>> disambiguated_operands;
   for (size_t i = 0; i < rex_operator->size(); ++i) {
     auto operand = rex_operator->getOperand(i);
-    if (dynamic_cast<const RexSubQuery*>(operand)) {
-      disambiguated_operands.emplace_back(rex_operator->getOperandAndRelease(i));
-    } else {
-      disambiguated_operands.emplace_back(disambiguate_rex(operand, ra_output));
-    }
+    // if (dynamic_cast<const RexSubQuery*>(operand)) {
```
[the remark this answers is not in the document]
Comment (ienkovich) [reply QueryEngine/RelAlgDagBuilder.cpp:1209]: This part is revised in the next part of changes.

File: QueryEngine/RelAlgDagBuilder.cpp [hunk 4]
```diff
@@ -1076,6 +1109,81 @@ std::unique_ptr<const RexScalar> parse_scalar_expr(const rapidjson::Value& expr,
...
+  CHECK(expr.IsObject());
+  if (expr.IsObject() && expr.HasMember("input")) {
+    return parse_input_expr(expr, ra_output);
+  }
+  if (expr.IsObject() && expr.HasMember("literal")) {
+    return parse_literal_expr(expr);
+  }
+  if (expr.IsObject() && expr.HasMember("op")) {
+    /*
+    const auto op_str = json_str(field(expr, "op"));
```
[the remark this answers is not in the document]
Comment (ienkovich) [reply QueryEngine/RelAlgDagBuilder.cpp:1167]: This part is revised in the next part of changes.

**Review by kurapov-peter — APPROVED** [review]
LGTM

**Pull request merged**